storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala

```diff
@@ -61,10 +66,58 @@ object MergeableStore {
       }
 
     /**
-     * Combine original keys with result after put. Some keys might have dropped,
-     * fill those using missingfn.
+     * Combine original keys with result after put.
      */
-    FutureOps.liftFutureValues(keySet, mPutResultFut, missingfn)
+    FutureOps.liftFutureValues(keySet, mPutResultFut, getFailures)
+  }
+
+  /**
+   * Collects keyed futures, partitioning out the failures.
+   */
+  private[algebra] def collectWithFailures[K, V](fs:Map[K, Future[V]]): (Future[Seq[(K, V)]], Map[K, Future[Nothing]]) = {
+    if (fs.isEmpty) {
+      (Future.value(Seq.empty[(K, V)]), Map.empty[K, Future[Nothing]])
+    } else {
+      val fsSize = fs.size
+      val results = new AtomicReferenceArray[Either[(K,Future[Nothing]), (K, V)]](fsSize)
+      val countdown = new AtomicInteger(fsSize)
+      val successCount = new AtomicInteger(0)
+      val pSuccess = new Promise[Seq[(K, V)]]
+      var failures = Map.empty[K, Future[Nothing]]
+
+      def collectResults() = {
+        if (countdown.decrementAndGet() == 0) {
+          val successArray = new Array[(K, V)](successCount.get())
+          var si = 0
+          var ri = 0
+          while (ri < fsSize) {
+            results.get(ri) match {
+              case Right(kv) =>
+                successArray(si) = kv
+                si += 1
+              case Left(kv) =>
+                failures = failures + kv
+            }
+            ri += 1
+          }
+          pSuccess.setValue(successArray)
+        }
+      }
+
+      for (((k, f), i) <- fs.iterator.zipWithIndex) {
+        f respond {
+          case Return(v) =>
+            results.set(i, Right(k -> v))
+            successCount.incrementAndGet()
+            collectResults()
+          case t@Throw(cause) =>
+            val failure = k -> Future.const(Throw(cause))
+            results.set(i, Left(failure))
+            collectResults()
+        }
+      }
+      (pSuccess, failures)
```

Comment:
failures has to be a future right? You are returning a copy here, but the futures are not done yet. If you could have a `Map` you could see what keys have failed, and you can't can you? I think this has to be `Future[Map[K, ..]]`. I am not sure `Future[Nothing]` is a better type that `Throwable` though.